### PR TITLE
Update and extension of Adding push constants support #1369

### PR DESCRIPTION
**Pull request by mtsr**
This extends @Neo-Zhixing 's #1369 with shader-reflection and a (albeit contrived) example.

@Neo-Zhixing I hope you don't mind. I left your authorship of the rebased commits intact.

I'd be particularly interested in feedback on how to make a more useful or simple example. But since `set_push_constants` requires DrawContext, I'm not sure if there's a real alternative.

### CODE REVIEW

**Comment by Neo-Zhixing** [comment]
This is really nice. Thank you for keeping my authorship.

**Comment by mtsr** [comment]
Because it's a good example I want to repeat the links from the other PR here:

> https://github.com/Neo-Zhixing/bevy_sky/blob/b49ac0520c751ab47936126fbeef828b44540c40/src/sky_node.rs#L139
> 
> https://github.com/Neo-Zhixing/bevy_sky/blob/b49ac0520c751ab47936126fbeef828b44540c40/src/procsky.frag#L10
> 
> https://github.com/Neo-Zhixing/bevy_sky/blob/b49ac0520c751ab47936126fbeef828b44540c40/examples/procedural_sky.rs#L8

**Comment by mtsr** [comment]
> Push constants would be user-facing in the sense that users could use them (like you are using them in the linked `bevy_sky` examples). For internal adoption, my only major concern is "web compatibility". I don't want to maintain two render paths yet (one for web and one for "native"). Imo we should have a real "motivating" use case before merging. If it can't be internal (for now), then I think it needs to be "allowing users to adopt push constants". And if thats true, we should probably have an example.

I'm looking at using push_constants for shadowmapping. That way I can pass the index of the pointlight and the face that needs to be rendered to the shadow map shader.

**Comment by mtsr** [comment]
Although the limited compatibility concerns me too. I'll look into pushing that data into an aligned uniform as an alternative.

File: crates/bevy_render/src/shader/shader_reflect.rs [hunk 1]
```diff
@@ -28,6 +28,22 @@ impl ShaderLayout {
                     bind_groups.push(bind_group);
                 }
 
+                let mut push_constant_ranges = Vec::new();
+                for push_constant_block in module.enumerate_push_constant_blocks(None).unwrap() {
+                    let range = push_constant_block.offset..push_constant_block.size;
+                    let mut stages = BindingShaderStage::NONE;
+                    if shader_stage.contains(ReflectShaderStageFlags::VERTEX) {
```
Comment (jondo2010) [inline crates/bevy_render/src/shader/shader_reflect.rs:35]: This seems like it could be potentially non-exhaustive. Room for improved error handling?

**Comment by mtsr** [comment]
@Neo-Zhixing Btw, if you're interested in putting in the work to get this approved, feel free to pull (some of) my commits into your PR and we can close this instead.

**Pull request closed without merging**